### PR TITLE
Add an option to dump the results of the tests to a JSON machine-readable file.

### PR DESCRIPTION
**Pull request by ehlemur-zz**
Add an option to dump the results of the tests to a JSON machine-readable file.
The format of the file is described in https://www.chromium.org/developers/the-json-test-results-format

### CODE REVIEW

**Comment by ehlemur-zz** [comment]
@kwiberg PTAL

**Comment by ehlemur-zz** [comment]
@pbos

File: gtest-parallel [hunk 1]
```diff
@@ -424,4 +470,6 @@ if options.print_test_times:
               if times.get_test_time(test_binary, test) is not None)
   for (time_ms, test_binary, test) in ts:
     print "%8s %s" % ("%dms" % time_ms, test)
+if options.dump_json_test_results:
+  test_results.dump_to_file(options.dump_json_test_results)
```
Comment (pbos) [inline gtest-parallel]: Can you hold on to this file handle for the duration of the tests? CollectTestResults() could take dump_json_test_results as a parameter and .dump_to_file could be .write() or something (and the no-op test_results could have an empty .write()).

Comment (pbos) [reply gtest-parallel]: Or call it .close() on CollectTestResults, and it writes on close, that makes sense to me. @kwiberg probably has a better method name, but I think CollectTestResults should hold onto the file handle.

Comment (ehlemur-zz) [reply gtest-parallel]: What about dump_to_file_and_close()?

**Comment by ehlemur-zz** [comment]
PTAL :)

File: gtest-parallel [hunk 2]
```diff
@@ -424,4 +472,6 @@ if options.print_test_times:
               if times.get_test_time(test_binary, test) is not None)
   for (time_ms, test_binary, test) in ts:
     print "%8s %s" % ("%dms" % time_ms, test)
+if options.dump_json_test_results:
+  test_results.dump_to_file_and_close()
```
Comment (pbos) [inline gtest-parallel]: Implement this in the IgnoreTestResults() class too and make it unconditional.

Also add an empty line before this one.

File: gtest-parallel [hunk 3]
```diff
@@ -295,6 +334,10 @@ if not (0 <= options.shard_index < options.shard_count):
                "(less than the number of shards)." %
                (options.shard_index, options.shard_count - 1))
 
+test_results = IgnoreTestResults()
+if options.dump_json_test_results:
+  test_results = CollectTestResults(options.dump_json_test_results)
```
Comment (pbos) [inline gtest-parallel]: I think this was better before as CTR(arg) if opt.dump... else ITR().

File: gtest-parallel [hunk 4]
```diff
@@ -186,6 +187,40 @@ class RawFormat:
...
+        results = results.setdefault(name, {})
+      results.update(result)
+
+  def dump_to_file_and_close(self):
+    json.dump(self.test_results, self.json_dump_file)
+    self.json_dump_file.close()
+
+class IgnoreTestResults(object):
+  def log(self, test, result):
+    pass
```
Comment (pbos) [inline gtest-parallel:222]: Add dump_to_file_and_close(self)

**Review by pbos — CHANGES_REQUESTED** [review]
dump_to_file_and_close sounds great

File: gtest-parallel [hunk 5]
```diff
@@ -334,9 +336,8 @@ if not (0 <= options.shard_index < options.shard_count):
                "(less than the number of shards)." %
                (options.shard_index, options.shard_count - 1))
 
-test_results = IgnoreTestResults()
-if options.dump_json_test_results:
-  test_results = CollectTestResults(options.dump_json_test_results)
+test_results = (IgnoreTestResults() if options.dump_json_test_results is None
+                else CollectTestResults(options.dump_json_test_results))
```
Comment (ehlemur-zz) [inline gtest-parallel:340]: What about like this, so the line doesn't get too long?

**Review by pbos — APPROVED** [review]

**Comment by pbos** [comment]
I want @kwiberg to OK it as well, then we can squash this. :)

Thanks!

**Review by kwiberg — APPROVED** [review]
Looks good to me!

**Pull request merged**